sdk/src/Services/S3/Custom/Transfer/Internal/MultipartDownloadCoordinator.cs

```diff
@@ -0,0 +1,499 @@
+/*******************************************************************************
+ *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use
+ *  this file except in compliance with the License. A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file.
+ *  This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ *  CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ *  specific language governing permissions and limitations under the License.
+ * *****************************************************************************
+ *    __  _    _  ___
+ *   (  )( \/\/ )/ __)
+ *   /__\ \    / \__ \
+ *  (_)(_) \/\/  (___/
+ *
+ *  AWS SDK for .NET
+ *  API Version: 2006-03-01
+ *
+ */
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Amazon.Runtime;
+using Amazon.S3.Model;
+
+namespace Amazon.S3.Transfer.Internal
+{
+    /// <summary>
+    /// Coordinates multipart downloads using PART or RANGE strategies per SEP spec.
+    /// Handles discovery, concurrent downloads.
+    /// </summary>
+    internal class MultipartDownloadCoordinator : IDownloadCoordinator
+    {
+        private readonly IAmazonS3 _s3Client;
+        private readonly BaseDownloadRequest _request;
+        private readonly DownloadCoordinatorConfiguration _config;
+        private readonly IPartDataHandler _dataHandler;
+        private readonly SemaphoreSlim _httpConcurrencySlots;
+        private readonly RequestEventHandler _requestEventHandler;
+        
+        private Exception _downloadException;
+        private bool _disposed = false;
+        private bool _discoveryCompleted = false;
+        private readonly object _lockObject = new object();
```

Comment:
This is an internal class only called `BufferedMultipartStream.InitializeAsync`. Only a single thread will go through this class so you don't need to add locking in this class.